src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                            {
+                                var resultFromOperationBlockAnalysis = ReachingDefinitionsAnalysis.Run(operationBlock, context.OwningSymbol, context.CancellationToken);
+                                if (!resultFromOperationBlockAnalysis.HasUnusedDefinitions())
+                                {
+                                    // Assert that even slow pass (dataflow analysis) would have yielded no unused definitions.
+                                    Debug.Assert(!ReachingDefinitionsAnalysis.Run(context.GetControlFlowGraph(operationBlock), context.OwningSymbol, context.CancellationToken)
+                                                 .HasUnusedDefinitions());
+                                    hasBlockWithAllUsedDefinitions = true;
+                                    continue;
+                                }
+                            }
+
+                            // Now perform the slower, precise, CFG based reaching definitions dataflow analysis to identify the actual unused definitions.
+                            var cfg = context.GetControlFlowGraph(operationBlock);
+                            var resultFromFlowAnalysis = ReachingDefinitionsAnalysis.Run(cfg, context.OwningSymbol, context.CancellationToken);
+                            resultsFromFlowAnalysis.Add(resultFromFlowAnalysis);
+
+                            foreach (var (unusedSymbol, unusedDefinition) in resultFromFlowAnalysis.GetUnusedDefinitions())
+                            {
+                                if (unusedDefinition == null)
+                                {
+                                    // Null definition indicates initial definition for parameter from method declaration.
+                                    // Report unused local function parameters (which are specific to this operation block) here.
+                                    // We process parameters for the context's owning symbol after analyzing all operation blocks for this symbol
+                                    // as we need to verify parameter usages across all operation blocks before flagging it as unused.
+                                    // For example, a constructor with both a constructor initializer and body will have two different operation blocks.
+                                    var unusedParameter = (IParameterSymbol)unusedSymbol;
+                                    if (isComputingUnusedParams &&
+                                        unusedParameter.ContainingSymbol.IsLocalFunction())
+                                    {
+                                        var hasReference = resultFromFlowAnalysis.SymbolsRead.Contains(unusedParameter);
+                                        _symbolStartAnalyzer.ReportUnusedParameterDiagnostic(unusedParameter, hasReference, context.ReportDiagnostic, context.Options, context.CancellationToken);
```

Comment:
hit a complexity wall here and gave up.  moving on.  

My recommendation would be to break some of these humongous methods out into many smaller mmethods.  Or, more ideally, use the 'method object' refactoring pattern.  Then, in that method-object, break things out such that hte topmost method is a very clear set of steps outlining what is going on, and each extracted method has a very small responsibility that can be stated clearly.  Right now, there is so much going on in this method that i've completely lost the plot.

Another way to think about this is the whole "the brain can keep track of like 7 things at a time" philosophy. :)  So, at any method level it's good to try to write such that the intend of that method itself is clear, with it responsible for just a few things.  Then if there is more compelxity needed for any of those steps, it is encapsulated in that step where you can pop-in/out as necessary.

**Importantly** the relation between the steps should be explicit and not implicit.  Data should not flow between the steps implicitly (esp. where mutation is concerned).

Reply:
Refactored with more comments and smaller concise functions.

---
In reply to: [230605710](https://github.com/dotnet/roslyn/pull/30777#discussion_r230605710) [](ancestors = 230605710)